### ISSUE TITLE
AccessViolationException

### ISSUE THREAD

**Issue by jaerith** [stellarbear/YaraSharp#10]
I'm attempting to use the library via a simple example:

```
YSInstance YSInstance = new YSInstance();

Dictionary<string, object> CONST_YS_EXTERNAL_VALUES = 
            new Dictionary<string, object>()
                {
                    { "filename", string.Empty },
                    { "filepath", string.Empty },
                    { "extension", string.Empty }
                };

//	Get list of YARA rules
List<string> ruleFilenames = 
                    Directory.GetFiles(@"F:\Configuration\Yara", "*.yara", SearchOption.AllDirectories).ToList();

YSRules reviewRules = null;

using (YSContext context = new YSContext())
{
    //	Compiling rules
    using (YSCompiler compiler = YSInstance.CompileFromFiles(ruleFilenames, CONST_YS_EXTERNAL_VALUES))
    {
        //  Get compiled rules
        reviewRules = compiler.GetRules();

        YSReport errors = compiler.GetErrors();
        YSReport warnings = compiler.GetWarnings();
    }
}

string[] list = new string[] { "Flare", "Something", "blah blah", "offensive" };
foreach (var TmpVal in list)
{
    byte[] commTextBytes = ASCIIEncoding.ASCII.GetBytes(TmpVal);

    List<YSMatches> Matches =
        YSInstance.ScanMemory(commTextBytes,
                                                  poReviewRules,
                                                  CONST_YS_EXTERNAL_VALUES,
                                                  1000);
}
```

But when I call "ScanMemory()", I get the following result:

**System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'**

Have you seen this before?

**Comment by ChrisDavies-MSFT** [comment]
Off the top of my head, I can’t see what’s wrong with that, but I haven’t done any debugging yet. I am a bit rammed right now, but I’ll try to take a look at it a bit later today.

Btw, I have a private branch I run here, which fixes up the timeout support and a few other extra scan methods. I’ll try to push those changes back shortly. Also, if you are not planning to pass external arguments, you can just pass null.

I’ll try to get back to you asap 😊

Chris Davies (MSTIC)
Senior Software Engineer
United Kingdom
TI - Science
Mobile: +447960650292
christopher.davies@microsoft.com<mailto:christopher.davies@microsoft.com>


[MSFT_logo]

**Comment by jaerith** [comment]
No worries.  Thanks for the quick response!

I should mention that I'm a newb with the Yara stuff, and I was just trying different libraries, in order to get a feel for what's out there.  In truth, I haven't installed Yara on my system yet or built the C++ library yet.  I just pulled down the Nuget package and started calling the APIs to become more familiar with it.  In fact,  I assume that the AccessViolationException could be a result of not having the actual Yara library/executable installed on my machine.  So, there's a chance that this is all my fault.  :)

**Comment by AdamWhiteHat** [comment]
> Have you seen this before?

Yeah, you're calling ScanMemory outside of your using statement: `using (YSContext context = new YSContext()) { ... }`

Your YSContext is already disposed of by then!

Try moving `YSInstance.ScanMemory(...)` to inside the curly braces.

**Comment by ChrisDavies-MSFT** [comment]
Yup, that's the answer! I'm sorry I never came back to this! FYI, at some point I plan to start a new repo and upgrade to V4. Unfortunately, StellarBear is no longer working on this project :(

**Comment by lanyouqi** [comment]
I plan to use C# FileSystemWatcher to write an automatic directory monitoring program. When a new file is created, the callback function uses Yara to review it. However, the Yara part needs to be run in the using (YSContext context = new YSContext()){} block. Doesn't that mean that every time a file is created, a compiler must be compiled? Is there any good solution? Thank you very much (from Google Translate)

private List<YSMatches> ScanFileWithYara(string filePath)
{
    try
    {
        using (YSContext context = new YSContext())
        {
            //	Compiling rules
            YSCompiler compiler = _instance.CompileFromFiles(ruleFilenames, externals);

            //  Get compiled rules
            YSRules rules = compiler.GetRules();

            //  Get errors
            YSReport errors = compiler.GetErrors();
            //  Get warnings
            YSReport warnings = compiler.GetWarnings();


            //  Some file to test yara rules
            string Filename = Path.GetFileName(filePath);
            Console.WriteLine(Alphaleonis.Win32.Filesystem.Path.GetFileName(Filename));
            Console.WriteLine(Alphaleonis.Win32.Filesystem.Path.GetFullPath(Filename));
            Console.WriteLine(Alphaleonis.Win32.Filesystem.Path.GetExtension(Filename));

            List<YSMatches> Matches = _instance.ScanFile(filePath, rules, new Dictionary<string, object>()
                {
                        { "filename", Path.GetFileName(filePath)},
                        { "filepath", Path.GetFullPath(filePath)},
                        { "extension", Path.GetExtension(filePath)}
                 }, 0);
            return Matches;

        }
    }
    catch (Exception ex)
    {
        Console.WriteLine($"[FSA] 扫描文件 {filePath} 时发生错误: {ex.Message}");
        return new List<YSMatches>();
    }

[code.txt](https://github.com/user-attachments/files/18923900/code.txt)

There seems to be some problem with the code format of giihub

**Comment by AdamWhiteHat** [comment]
> Doesn't that mean that every time a file is created, a compiler must be compiled? Is there any good solution?

Not necessarily. I have used YaraSharp in a long-running application before, I just write a wrapper class to contain the Yara objects, and inherit from IDisposable, calling dispose on the Yara objects in the Dispose method. (Example below.)

The way the compiler works, I compile related bundles/sets of rules together, dispose of the compiler, but hold on to the compiled object. I usually set this up in advance of any scanning. Then when there is a file to scan, run the list of compiled rule sets over them, reporting what each rule set found, if anything.

Although I hold on to the compiled objects and make sure I dispose of them, but since they are (probably) living for the lifetime of the application, if you forgot about disposing of them, it probably wouldnt matter much anyways, as the application is about to be closed anyways, and any managed objects will get garbage collected when the application terminates. 

The only place where you might want to watch out is if this all lives in a library, and you load the library up manually using LoadAssembly into a variable and pass a references of that variable instance around to other libraies/processes/appdomains that are left running; Then maybe it gets held in memory longer than you anticipated.


Below is a class that shows an example of how I use YaraSharp in a long-running scenario. I copied and pasted code from a project  that uses YaraSharp into a class, thereby distilling down all the relevant bits from all the irrelevant bits from the rest of the application. 

```csharp
public class YaraWrapper : IDisposable
{
    public bool IsDisposed { get; private set; }

    private List<YSRules> _rulesToCleanup;
    private List<YSScanner> _yaraScanners;
    private YSInstance _yaraInstance;
    private YSContext _yaraContext;

    public YaraWrapper()
    {
        IsDisposed = false;
        _rulesToCleanup = new List<YSRules>();
        _yaraScanners = new List<YSScanner>();
        _yaraInstance = new YSInstance();
        _yaraContext = new YSContext();
    }

    public void AddRuleSet(List<string> yaraRuleFiles)
    {
        YSRules compiledRules = null;

        using (YSCompiler compiler = _yaraInstance.CompileFromFiles(yaraRuleFiles, null))
        {

            YSReport warnings = compiler.GetWarnings();
            if (!warnings.IsEmpty())
            {
                // Report warnings
            }
            YSReport errors = compiler.GetErrors();
            if (!errors.IsEmpty())
            {
                // Log errors. Maybe bail?
            }

            compiledRules = compiler.GetRules();
            _rulesToCleanup.Add(compiledRules);
        }

        YSScanner scanner = new YSScanner(compiledRules, null);
        _yaraScanners.Add(scanner);

        // Personally, I use a Dictionary<string, YSScanner> here, where the key is the SHA256 hash of the rule files
        // Then for any rule set, I get the hash and check the dictionary, thereby caching any sets of rules,
        // but this may be overkill for your needs.
        // Just a thought.
    }

    public List<string> ScanFile(string fullFilePath)
    {
        List<string> results = new List<string>();

        foreach (YSScanner scanner in _yaraScanners)
        {
            List<YSMatches> scanResults = scanner.ScanFile(filePath);
            if (scanResults.Any())
            {
                results = scanResults.Select(match => match.Rule.Identifier).ToList();
                // OR 
                // List<YSMatch> matches =  scanResults.SelectMany(match => match.Matches.SelectMany(kvp => kvp.Value)).ToList();
            }
        }

        return results.Distinct();
    }

    public void Dispose()
    {
        if (!IsDisposed)
        {
            IsDisposed = true;

            foreach (YSRules rule in _rulesToCleanup)
            {
                rule.Dispose();
            }
            foreach (YSScanner scanner in _yaraScanners)
            {
                scanner.Dispose();
            }

            _yaraContext.Dispose();
            _yaraInstance = null;
        }
    }
}
```

 If for some reason you want to see how I'm actually using it in an application, with full context and everything, search for Yara or "YS" in my project that uses it, [Judge, Jury and Executable](https://github.com/AdamWhiteHat/Judge-Jury-and-Executable).

**Comment by lanyouqi** [comment]
> > 这难道不是意味着每次创建文件时都必须编译编译器吗？有什么好的解决办法吗？
> 
> 不一定。我以前在长时间运行的应用程序中使用过 YaraSharp，我只是编写一个包装类来包含 Yara 对象，并从 IDisposable 继承，在 Dispose 方法中对 Yara 对象调用 dispose。（示例如下。
> 
> 编译器的工作方式是，我将相关的 bundles/sets of rules 编译在一起，丢弃编译器，但保留编译后的对象。我通常在任何扫描之前进行设置。然后，当有要扫描的文件时，对它们运行已编译规则集的列表，报告每个规则集找到的内容（如果有）。
> 
> 虽然我保留了已编译的对象并确保我处理了它们，但是由于它们（可能）在应用程序的生命周期内都存在，如果您忘记处理它们，那么这可能也无关紧要，因为应用程序无论如何都即将关闭，并且当应用程序终止时，任何托管对象都会被垃圾回收。
> 
> 唯一可能需要注意的地方是，如果这一切都位于库中，并且您使用 LoadAssembly 手动将库加载到变量中，并将该变量实例的引用传递给其他保持运行的库/进程/appdomains;那么它可能在内存中的保存时间比您预期的要长。
> 
> 下面是一个类，它显示了我如何在长时间运行的方案中使用 YaraSharp 的示例。我将使用 YaraSharp 的项目中的代码复制并粘贴到一个类中，从而从应用程序其余部分的所有不相关位中提炼出所有相关位。
> 
> public class YaraWrapper : IDisposable
> {
>     public bool IsDisposed { get; private set; }
> 
>     private List<YSRules> _rulesToCleanup;
>     private List<YSScanner> _yaraScanners;
>     private YSInstance _yaraInstance;
>     private YSContext _yaraContext;
> 
>     public YaraWrapper()
>     {
>         IsDisposed = false;
>         _rulesToCleanup = new List<YSRules>();
>         _yaraScanners = new List<YSScanner>();
>         _yaraInstance = new YSInstance();
>         _yaraContext = new YSContext();
>     }
> 
>     public void AddRuleSet(List<string> yaraRuleFiles)
>     {
>         YSRules compiledRules = null;
> 
>         using (YSCompiler compiler = _yaraInstance.CompileFromFiles(yaraRuleFiles, null))
>         {
> 
>             YSReport warnings = compiler.GetWarnings();
>             if (!warnings.IsEmpty())
>             {
>                 // Report warnings
>             }
>             YSReport errors = compiler.GetErrors();
>             if (!errors.IsEmpty())
>             {
>                 // Log errors. Maybe bail?
>             }
> 
>             compiledRules = compiler.GetRules();
>             _rulesToCleanup.Add(compiledRules);
>         }
> 
>         YSScanner scanner = new YSScanner(compiledRules, null);
>         _yaraScanners.Add(scanner);
> 
>         // Personally, I use a Dictionary<string, YSScanner> here, where the key is the SHA256 hash of the rule files
>         // Then for any rule set, I get the hash and check the dictionary, thereby caching any sets of rules,
>         // but this may be overkill for your needs.
>         // Just a thought.
>     }
> 
>     public List<string> ScanFile(string fullFilePath)
>     {
>         List<string> results = new List<string>();
> 
>         foreach (YSScanner scanner in _yaraScanners)
>         {
>             List<YSMatches> scanResults = scanner.ScanFile(filePath);
>             if (scanResults.Any())
>             {
>                 results = scanResults.Select(match => match.Rule.Identifier).ToList();
>                 // OR 
>                 // List<YSMatch> matches =  scanResults.SelectMany(match => match.Matches.SelectMany(kvp => kvp.Value)).ToList();
>             }
>         }
> 
>         return results.Distinct();
>     }
> 
>     public void Dispose()
>     {
>         if (!IsDisposed)
>         {
>             IsDisposed = true;
> 
>             foreach (YSRules rule in _rulesToCleanup)
>             {
>                 rule.Dispose();
>             }
>             foreach (YSScanner scanner in _yaraScanners)
>             {
>                 scanner.Dispose();
>             }
> 
>             _yaraContext.Dispose();
>             _yaraInstance = null;
>         }
>     }
> }
> 如果出于某种原因，您想了解我如何在应用程序中实际使用它，以及完整的上下文和所有内容，请在使用它的项目中搜索 Yara 或“YS”，[Judge、Jury 和 Executable](https://github.com/AdamWhiteHat/Judge-Jury-and-Executable)。

Thank you very much, it's a great method. I have successfully adopted it
